libs/ngx-sherfire/src/lib/firestore/derivable-firestore.service.ts

```diff
@@ -0,0 +1,136 @@
+import { Injectable, NgZone } from '@angular/core';
+import { Derivable, DerivableAtom, SettableDerivable, error, lens } from '@skunkteam/sherlock';
+import { DerivableCacheOptions, derivableCache, fromEventPattern } from '@skunkteam/sherlock-utils';
+import {
+    CollectionReference,
+    DocumentData,
+    DocumentReference,
+    DocumentSnapshot,
+    FirestoreError,
+    Query,
+    QueryDocumentSnapshot,
+    QuerySnapshot,
+    onSnapshot,
+    setDoc,
+} from 'firebase/firestore';
+import { NgxSherfireModule } from '../ngx-sherfire.module';
+
+@Injectable({ providedIn: NgxSherfireModule })
+export class DerivableFirestore {
+    constructor(private readonly zone: NgZone) {}
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of updated QuerySnapshots that contain all
+     * QueryDocumentSnapshots matching the query.
+     */
+    snapshot$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<QuerySnapshot<App, Db>>;
+    /**
+     * Build a derivable based on a DocumentReference that returns a stream of updated DocumentSnapshots.
+     */
+    snapshot$<App, Db extends DocumentData>(ref: DocumentReference<App, Db>): Derivable<DocumentSnapshot<App, Db>>;
+    snapshot$<App, Db extends DocumentData>(
+        v: DocumentReference<App, Db> | Query<App, Db>,
+    ): Derivable<DocumentSnapshot<App, Db>> | Derivable<QuerySnapshot<App, Db>> {
+        return v instanceof DocumentReference
+            ? this.CACHES.doc.snapshot(v)
+            : v instanceof CollectionReference
+            ? this.CACHES.collection.snapshot(v)
+            : this.querySnapshot$(v);
+    }
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of updates with an array of QueryDocumentSnapshot.
+     */
+    docs$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<Array<QueryDocumentSnapshot<App, Db>>> {
+        return query instanceof CollectionReference ? this.CACHES.collection.docs(query) : this.queryDocs$(query);
+    }
+
+    /**
+     * Build a derivable based on a Query or CollectionReference that returns a stream of arrays of document contents.
+     */
+    data$<App, Db extends DocumentData>(query: Query<App, Db>): Derivable<App[]>;
+    /**
+     * Build a derivable based on a DocumentReference that returns a stream of document contents. The returned derivable is settable and
+     * setting a value to the derivable will update the firestore document.
+     */
+    data$<App, Db extends DocumentData>(ref: DocumentReference<App, Db>): SettableDerivable<App>;
+    data$<App, Db extends DocumentData>(
+        v: DocumentReference<App, Db> | Query<App, Db>,
+    ): SettableDerivable<App> | Derivable<App[]> {
+        return v instanceof DocumentReference
+            ? this.CACHES.doc.data(v)
+            : v instanceof CollectionReference
```

Comment:
Testing for `CollectionReference` (or `Query`) used to be with `'path' in this` in `augmentQueries` but I feel that we can now safely use instanceof.